apiserver/client/status.go

```diff
@@ -533,21 +635,50 @@ func makeMachineStatus(machine *state.Machine) (status params.MachineStatus) {
 		}
 		status.DNSName = addr.Value
 
-		mAddrs := machine.Addresses()
-		if len(mAddrs) == 0 {
-			logger.Debugf("no IP addresses fetched for machine %q", instid)
-			// At least give it the newly created DNSName address, if it exists.
-			if addr.Value != "" {
-				mAddrs = append(mAddrs, addr)
+		status.NetworkInterfaces = make(map[string]params.NetworkInterface, len(linkLayerDevices))
+		for _, llDev := range linkLayerDevices {
+			device := llDev.Name()
+			ips := []string{}
+			gw := []string{}
+			ns := []string{}
+			sp := make(set.Strings)
+			for _, ipAddress := range ipAddresses {
+				if ipAddress.DeviceName() == device {
+					ips = append(ips, ipAddress.Value())
+					// We don't expect to find more than
+					// one ipAddress on a device with a
+					// list of nameservers, but append in
+					// any case.
+					if len(ipAddress.DNSServers()) > 0 {
+						ns = append(ns, ipAddress.DNSServers()...)
+					}
+					// There should only be one gateway per
+					// device (per machine, in fact, as we
+					// don't store metrics). If we find
+					// more than one we should show them
+					// all.
+					if ipAddress.GatewayAddress() != "" {
+						gw = append(gw, ipAddress.GatewayAddress())
+					}
+					// There should only be one space per
+					// address, but it's technically
+					// possible to have more than one
+					// address on an interface. If we find
+					// that happens, we need to show all
+					// spaces, to be safe.
+					sp = spaces[device]
+				}
 			}
-		}
-		for _, mAddr := range mAddrs {
-			switch mAddr.Scope {
-			case network.ScopeMachineLocal, network.ScopeLinkLocal:
-				continue
+			status.NetworkInterfaces[device] = params.NetworkInterface{
+				IPAddresses:    ips,
+				MACAddress:     llDev.MACAddress(),
+				Gateway:        strings.Join(gw, " "),
```

Comment:
I didn't like the idea of the output in an incorrect type for the expected data, but instead captured any erroneous extras at formatting (as noted in the comments). Obviously, you could show an error if you get what you don't expect, but I erred on the side of showing the operator any multiple gateways (since this is OK in real life, with metrics, but not in Juju) and any multiple spaces. This seemed the most useful for troubleshooting.